samples/java-protobuf-tracing/src/main/java/com/example/ControllerAction.java

```diff
@@ -0,0 +1,94 @@
+package com.example;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Charsets;
+import com.google.protobuf.Empty;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Scope;
+import kalix.javasdk.action.ActionCreationContext;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.Charset;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public class ControllerAction extends AbstractControllerAction {
+
+  String url = "https://jsonplaceholder.typicode.com/posts/1";
+  HttpClient httpClient = HttpClient.newHttpClient();
+  public ControllerAction(ActionCreationContext creationContext) {}
+
+
+  @Override
+  public Effect<ControllerActionApi.MessageResponse> callAsyncEndpoint(Empty empty) {
+      // tag::get-tracer[]
+      Optional<Tracer> tracerOpt = actionContext().getOpenTelemetryTracer();
+      // end::get-tracer[]
+      CompletableFuture<HttpResponse<Post>> futureResponse;
+
+    if(tracerOpt.isPresent()) {
+      futureResponse = callAsyncService(tracerOpt.get());
+    } else {
+      futureResponse = callAsyncService();
+    }
+    CompletionStage<ControllerActionApi.MessageResponse> asyncComputation = futureResponse.toCompletableFuture()
+            .thenApply(response -> {
+      return ControllerActionApi.MessageResponse.newBuilder().setMessage(response.body().title).build();
+    });
+    return effects().asyncReply(asyncComputation);
+  }
+
+
+  private CompletableFuture<HttpResponse<Post>> callAsyncService(){
+    HttpRequest httpRequest = HttpRequest.newBuilder()
+            .uri(URI.create(url))
+            .build();
+    //Async call to external service
+    return httpClient.sendAsync(httpRequest,
+            new JsonResponseHandler<>(Post.class));
+
+
+  }
+
+
+  //  tag::create-close-span[]
+  private CompletableFuture<HttpResponse<Post>> callAsyncService(Tracer tracer) {
+  // end::create-close-span[]
+
+    // tag::create-close-span[]
+    Span span  = tracer 
+            .spanBuilder("https://jsonplaceholder.typicode.com/posts/{}")
+            .setParent(actionContext().metadata().traceContext().asOpenTelemetryContext())// <1>
+            .startSpan(); // <2>
+    span.setAttribute("post", "1");// <3>
+
+    CompletableFuture<HttpResponse<Post>> responseFuture = callAsyncService();
+
+    try (Scope scope = span.makeCurrent()) {// <4>
```

Comment:
@franciscolopezsancho, maybe something more to watch out. 

We will close the scope and end the span as soon as the try/finally block is executed.

But the `thenAccept` and `exceptionally` will happen later, when the future completes. So I guess this will produce unpredictable results.

Reply:
So maybe this is way we didn't see the error marker in the span. Good idea. I'll take that into account.

Reply:
That was the case yes. We need to move the `span.end()` outside the `finally`, putting in both callbacks.

Reply:
What about the scope? Why are opening the scope in a try-with-resources? 

The resource is closed as soon as the try block ends. Are we possibly add the span on the wrong scope? 

Maybe the issue reported by Eduardo is similar to this. 
https://github.com/lightbend/kalix-runtime/issues/2369#issuecomment-2014894746

A span is added in the wrong scope because of some async execution order.